pkg/service/managedautoevents.go

```diff
@@ -0,0 +1,64 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2020 VMware
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package service
+
+import (
+	"fmt"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/autoevent"
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+// AddDeviceAutoEvent adds a new AutoEvent to the Device with given name
+func (s *DeviceService) AddDeviceAutoEvent(deviceName string, event contract.AutoEvent) error {
+	device, ok := cache.Devices().ForName(deviceName)
+	if !ok {
+		msg := fmt.Sprintf("Device %s cannot be found in cache", deviceName)
+		s.LoggingClient.Error(msg)
+		return fmt.Errorf(msg)
+	}
+
+	for _, e := range device.AutoEvents {
+		if e.Resource == event.Resource {
+			s.LoggingClient.Debug(fmt.Sprintf("Updating existing auto event %s for device %s\n", e.Resource, deviceName))
+			e.Frequency = event.Frequency
+			e.OnChange = event.OnChange
+			return nil
```

Comment:
suggest still call `RestartForDevice` in this case or refactor the executor.go otherwise the frequency update won't be applied.